### PR TITLE
Add new raw_components method

### LINKED ISSUE #2627: Make raw_components version of bokeh.embed.components
When I'm embedding Bokeh plots I currently need to do some workarounds so that:
* the components function does not wrap the generated bokeh javascript in a script tag
* i can have a list of the plot ids

I propose a raw_components method that works very like the components method, but instead of returning `(script, divs)` returns `(unwrapped_script, plot_ids)`.

### PR DESCRIPTION
**Pull request by birdsarah**
issues: fixes #2627

* includes one new test for components
* refactors some methods in embed for reuse by raw_components
* raw_components returns the un-tagged javascript and the raw plot_info dictionaries

### CODE REVIEW

**Comment by bryevdv** [comment]
I'm -1 on this change as-is. I don't like the name `raw_components` but more importantly the two functions are almost identical. I think this would be better as an optional switch to `components`:
```
components(..., wrap_script=True)
```
or something similar.

**Comment by birdsarah** [comment]
@bryevdv. Thanks for the feedback. Can you go a bit further though as there are two differences:
1 - As you've noted, the script is not wrapped.
2 - It does not return but the divs, but returns the plot information dictionaries.

Regarding the methods being the same, they both are and aren't, I wanted to dry out more, but couldn't find a way that was also still easy to understand what's going on.

**Comment by birdsarah** [comment]
Another thought for the direction of this PR would be to keep the re-factoring of components, but remove the raw_components method. That way I could do my own raw_components method when I need it without having to jump through so many hoops, but we don't have to build it in to bokeh yet. 

This may give more opportunity for experience of how both the non-wrapped script and the plot info dictionary is used.

**Comment by fpliger** [comment]
I +1 on the suggestion add parameters to `components` change the output instead of having a new `raw_components`. One specific thing I don't like much about this `raw_components` is that it ties the `raw_script` with un-tagged javascript and dict returning just the plots id (not the divs) but it's may not be always true.

**Comment by birdsarah** [comment]
@fpliger I'm afraid I really don't understand your comment, can you explain a bit more.

**Comment by birdsarah** [comment]
@fpliger - let me elaborate a bit more.

>  One specific thing I don't like much about this `raw_components` is that it ties the raw_script with un-tagged javascript and dict returning just the plots id (not the divs) but it's may not be always true. 

That's what you don't like about raw_components, but your describing exactly the feature I wanted. So I feel a bit stuck how to move forward.

What do you think of my suggestion to keep the refactor but remove the `raw_components` method

**Comment by fpliger** [comment]
@birdsarah sorry, not really in good shape (and probably shouldn't be trying to give meaningful feedbacks :-)

Anyway, if I got if right, on the proposal as is `raw_components` returns `(unwrapped_script, plot_ids)` and  `components` returns `(script, divs)`. FWIW the `script` shouldn't be tied to only one of `plot_ids` or `divs` at all. If instead of having `components` and `raw_components` we only have `components` that accepts arguments to specify [separately] if you want `script` or `unwrapped_script` and `plot_divs` or `plot_ids` then you have the same functionality you need with much more flexibility.... Unless I'm missing something.

**Comment by birdsarah** [comment]
Thanks @fpliger. Sorry you're not feeling well. @bryevdv was having a real brain fart about this and have just done a re-factor which really cleans things up. Am going to work on your suggestion of just updating the components method signature instead.

**Comment by damianavila** [comment]
+1 to @bryevdv and @fpliger comments...

File: bokeh/embed.py [hunk 1]
```diff
@@ -63,19 +64,43 @@ def components(plot_objects, resources=None):
 
         resources : Deprecated argument
 
+        wrap_script (boolean, optional): If set to false, the returned javascript is not wrapped
+        in a script tag. (default: True)
```
Comment (bryevdv) [inline bokeh/embed.py]: Maybe better expressed with "If True, ... (default: True)"

File: bokeh/embed.py [hunk 2]
```diff
@@ -63,19 +64,43 @@ def components(plot_objects, resources=None):
...
+
+        wrap_plot_info (boolean, optional): If true, then a set of divs are returned as described
+        above. If set to false, then dictionaries are returned that can be used to manually
+        build your own divs. The dictionary contains following information:
+        {
+            'modelid':  'The plots id, which can be used in the Bokeh.index',
+            'elementid': 'The css identifier the BokehJS will look for to target the plot',
+            'modeltype': 'Plot',
+        }
+        (default: True)
```
Comment (bryevdv) [inline bokeh/embed.py]: Docstrings can be multi-part, I like to put the default closer to the beginning:

```
        wrap_plot_info (boolean, optional): If true, returns divs as described above . If False, 
        then returns dictionaries that can be used to manually build your own divs. (default: True)

        Returned dictionaries contain the following information:
        {
            'modelid':  'The plots id, which can be used in the Bokeh.index',
            'elementid': 'The css identifier the BokehJS will look for to target the plot',
            'modeltype': 'Plot',
        }
```
Might need to check the format, it's possible the lines underneath have to be indented, I can't remember offhand.

File: bokeh/embed.py [hunk 3]
```diff
@@ -63,19 +64,43 @@ def components(plot_objects, resources=None):
...
-             'a future version.', DeprecationWarning, stacklevel=2)
+    all_models, plots, plot_info, divs = _get_components(plot_objects, resources)
+
+    if wrap_script:
+        script = _get_script(all_models, plots)
+    else:
+        script = _get_js(all_models, plots)
+    script = encode_utf8(script)
+
+    return script, divs if wrap_plot_info else plot_info
```
Comment (bryevdv) [inline bokeh/embed.py]: I typically try to prefer simpler expressions:
``` 
if wrap_plot_info: 
    return script, divs
else:
    return script, plot_info
```

**Comment by birdsarah** [comment]
Don't think including all this information in the args block is good. Just noticed the current documentation for components is messed up: 

http://bokeh.pydata.org/en/latest/docs/reference/resources_embedding.html#bokeh.embed.components

Working up a new docstring.

**Comment by birdsarah** [comment]
reformatted the docstring to the "Napoleon NumPy" style, and seems to be looking better.

image

**Comment by bryevdv** [comment]
Oh wow that `Returns` section is great I didn't realize you could do that. And yes, all docstrings should be in Google style (not Numpydoc style) for parsing by Sphinx-Napoleon

**Comment by bryevdv** [comment]
FYI we should merge into `0.10` branch right after dev build next week. There will probably be a few conflicts, I think the deprecated `resources` param has already been removed in `0.10`, e.g.

File: bokeh/embed.py [hunk 4]
```diff
@@ -28,27 +28,55 @@
...
+    Args
+    ----
+    plot_objects : PlotObject|list|dict|tuple
+        A single PlotObject, a list/tuple of PlotObjects, or a dictionary of keys and PlotObjects.
+
+    resources :
+        Deprecated argument
+
+    wrap_script : boolean, optional
+        If true, the returned javascript is wrapped in a script tag. (default: True)
```
Comment (bryevdv) [inline bokeh/embed.py]: capital T `True`

File: bokeh/embed.py [hunk 5]
```diff
@@ -28,27 +28,55 @@
...
+
+    wrap_plot_info : boolean, optional
+        If true, then a set of divs are returned.
```
Comment (bryevdv) [inline bokeh/embed.py]: ditto

File: bokeh/embed.py [hunk 6]
```diff
@@ -28,27 +28,55 @@
...
+        If set to false, then dictionaries are returned that can be used to manually
```
Comment (bryevdv) [inline bokeh/embed.py]: capital F `False`

File: bokeh/embed.py [hunk 7]
```diff
@@ -28,27 +28,55 @@
...
+        build your own divs. (default: True)
 
-        The following illustrates how different input types correlate to outputs:
+        If false, the returned dictionary contains the following information::
```
Comment (bryevdv) [inline bokeh/embed.py]: same

File: bokeh/embed.py [hunk 8]
```diff
@@ -28,27 +28,55 @@
...
+                'modeltype': 'The Bokeh model name e.g. Plot, Slider',
+            }
+
+
+
+    Returns
+    -------
+    (script, div[s]) :  UTF-8 encoded
+
+        The output, depends on the input as follows::
```
Comment (bryevdv) [inline bokeh/embed.py]: I don't think the comma is needed here

File: bokeh/embed.py [hunk 9]
```diff
@@ -59,23 +87,40 @@ def components(plot_objects, resources=None):
             components({"Plot 1": plot_1, "Plot 2": plot_2})
             #=> (script, {"Plot 1": plot_1_div, "Plot 2": plot_2_div})
 
-        An example can be found in examples/embed/embed_multiple.py
+    (raw_script, plot_info[s]) : UTF-8 encoded
 
-        resources : Deprecated argument
+        The output, depends on the input as follows::
```
Comment (bryevdv) [inline bokeh/embed.py]: or here

File: bokeh/embed.py [hunk 10]
```diff
@@ -59,23 +87,40 @@ def components(plot_objects, resources=None):
             components({"Plot 1": plot_1, "Plot 2": plot_2})
             #=> (script, {"Plot 1": plot_1_div, "Plot 2": plot_2_div})
```
Comment (bryevdv) [inline bokeh/embed.py]: I think `plot1_div`, etc. would be a little easier on the eyes. Would recommend that throughout (below for `plot2_dict` too, etc.)

File: bokeh/embed.py [hunk 11]
```diff
@@ -59,23 +87,40 @@ def components(plot_objects, resources=None):
...
-        An example can be found in examples/embed/embed_multiple.py
+    (raw_script, plot_info[s]) : UTF-8 encoded
 
-        resources : Deprecated argument
+        The output, depends on the input as follows::
 
-    Returns:
-        (script, div[s]): UTF-8 encoded
+            components(plot, wrap_script=False, wrap_plot_info=False)
+            #=> (javascript, plot_dict)
```
Comment (bryevdv) [inline bokeh/embed.py]: I think i like spaces after hashes, even here:
```
            components(plot, wrap_script=False, wrap_plot_info=False)
            # => (javascript, plot_dict)
```

**Comment by bryevdv** [comment]
OK, I made several small fiddly comments, documentation brings out the nitpicker in me :)

**Comment by birdsarah** [comment]
@bryevdv - any reason this is going in `0.10` not `master`? It doesn't break any current behavior, so I was thinking it could just straight in.

Thanks for the detailed doc review - details are great!

**Comment by bryevdv** [comment]
This is going into master. I meant that we need to bring `0.10` branch up to date with master too, sorry if that was not clear.

**Comment by birdsarah** [comment]
Ah....I see - yes now your comment makes a lot more sense - happy to help with that merge where I can.

**Comment by bryevdv** [comment]
@birdsarah FYI I think #2757 (fairly serious bug fix) will cause a merge conflict with this PR

**Comment by birdsarah** [comment]
No sweat - thanks for the headsup @bryevdv

**Comment by birdsarah** [comment]
@bryevdv, I've made the small documentation tweaks you requested, so marking this as ready again.

**Comment by birdsarah** [comment]
@bryevdv - have fixed up the merge conflict from #2757, this should be ready to go.

**Comment by damianavila** [comment]
OK, this LGTM and tested OK in some of the embed examples (not tested by TravsiCI)... let's get this merged now.

**Pull request merged**